### PR TITLE
minimize diff format by removing %{} useless parts. this drastically …

### PR DESCRIPTION
**Pull request by seb3s**
…improves component updates when a lot of them are involved.

### CODE REVIEW

**Comment by azazel75** [comment]
Maybe it could be a good idea to check/update the failing tests as well :stuck_out_tongue_winking_eye:

**Comment by seb3s** [comment]
Yes sure. These were expected to fail. Bu I prefer to wait for the feedback from josé & chris. I'm not sure they will really like the way I've optimized the diff (I removed parts I think are useless). I will then update the tests accordingly if it's ok.

**Comment by josevalim** [comment]
This is intentional. The front-end needs to know which components are involved in an update, even if their diffs are empty, because they can be transitive to another component update.

For example, LV -> C1 -> C2, C2 may have a non-empty diff, but C1 is empty, we can't delete C1 because the client needs to know this relationship.

We could stop sending this if the client is updating to keep this relationship, but I don't think it does today. See this commit for more info: 23333b9f120beeb2d859aee14d2438486a8f636e

**Comment by seb3s** [comment]
Hi Jose,
I've got nested components as well and all is working well in my app. Maybe I'm not facing the case you described. I would love to see a working example that is broken by my patch. I will keep it on my side as it results in an clean and responsive app versus non usable app for large sets in my case (trello style boards with lots of cards, all components). I wanted to push liveview a bit further to see if it can fit my needs for the future ;-)
Cheers,
Sébastien.

**Comment by josevalim** [comment]
Yeah, I started talking to Chris after this PR and it may be that we can apply parts of your PR. Some other parts would lead to broken behaviour though (for example, it seems it breaks rendering a component without assigns).

File: lib/phoenix_live_view/diff.ex [hunk 1]
```diff
@@ -465,9 +465,13 @@ defmodule Phoenix.LiveView.Diff do
...
         socket = Utils.clear_changed(%{socket | fingerprints: component_prints})
-        {socket, pending_components, Map.put(component_diffs, cid, diff), components}
+        if diff == %{} do
+          {socket, pending_components, component_diffs, components}
+        else
+          {socket, pending_components, Map.put(component_diffs, cid, diff), components}
+        end
       else
-        {socket, pending_components, Map.put(component_diffs, cid, %{}), components}
+        {socket, pending_components, component_diffs, components}
```
Comment (josevalim) [inline lib/phoenix_live_view/diff.ex:474]: To clarify, this change would be safe to apply. The one directly above isn't and the other one I am not so sure.

**Comment by seb3s** [comment]
Hi,
Can you precise what is a component without assigns?
I made some tests with:
<%= live_component(@socket, KandeskWeb.TestComponent) %> 
with or without :id (stateless vs stateful) and nothing breaks in my tests right now.

The log shows that for assigns:

[info] assigns: %{flash: %{}, id: nil, socket: #Phoenix.LiveView.Socket<assigns: %Phoenix.LiveView.Socket.AssignsNotInSocket{}, changed: %{flash: true, id: true}, endpoint: KandeskWeb.Endpoint, id: "phx-FhATXJ0BpMX4TC1C", parent_pid: nil, root_pid: #PID<0.4681.0>, router: KandeskWeb.Router, view: KandeskWeb.IndexLive, ...>}

which seems to be normal liveview stuff.

PS: my performance sensible tests are running with 1500+ components in the liveview, two levels nested (and one with more than 1000 components itself which is important for the test)

**Comment by josevalim** [comment]
Make sure that your test component receives an ID and does not set any assigns on `update`. Always returning the original socket with no changes.

**Comment by josevalim** [comment]
In any case, I find it unlikely that you are leveraging the code branch that has the bug, probably the biggest gain for you is the last change that you did on the file. The other two are probably very minor.

**Comment by seb3s** [comment]
No,
all the modifications I did have a huge impact on some use cases. I've got sorting and moving components from places to others. This results in nesting the components differently, so that lots of empty diffs are generated, which is killing the client rendering (thousands of updates to loop for no modifications at the end).

short extract =>
72: {1: {}, 3: {}}
73: {1: {}, 3: {}}
74: {1: {}, 3: {}}
75: {1: {}, 3: {}}
76: {1: {}, 3: {}}
77: {1: {}, 3: {}}
78: {1: {}, 3: {}}
79: {1: {}, 3: {}}
80: {1: {}, 3: {}}
81: {1: {}, 3: {}}
82: {1: {}, 3: {}}
83: {1: {}, 3: {}}
84: {1: {}, 3: {}}
85: {1: {}, 3: {}}

I even has another modification that I finally didn't put in the PR as It was not clean enough to me :-)

**Comment by josevalim** [comment]
I see. Good point about sorting. But your `diff == %{}` branch should at least check if the component is not new. If it is new we should always send the diff, otherwise the client has invalid data.

**Comment by seb3s** [comment]
Yep, adding !new is perfectly fine with my use case.

File: lib/phoenix_live_view/diff.ex [hunk 2]
```diff
@@ -293,10 +293,10 @@ defmodule Phoenix.LiveView.Diff do
...
         diff =
-          if serialized do
-            Map.put(diff, counter, serialized)
-          else
-            diff
+          case serialized do
+            nil -> diff
+            map when map == %{} -> diff
+            _ -> Map.put(diff, counter, serialized)
           end
```
Comment (josevalim) [inline lib/phoenix_live_view/diff.ex:300]: What is the scenario this optimization applies?

**Comment by seb3s** [comment]
My app is liveview (board) => components (columns) => components (tasks)
I've got tags defined at board level (shared everywhere) but use at task level (2 components level inside) for every tasks (1000+).
When I update the tags definition with the optimization =>
diff: {3: {,…}, 10: {d: [[1], [2], [3], [4], [5], [6], [7], [8], [9]]}, 11: "",…}
3: {,…}
10: {d: [[1], [2], [3], [4], [5], [6], [7], [8], [9]]}
11: ""
c: {1: {0: {9: {,…}}}, 2: {0: {9: {,…}}}, 3: {0: {9: {,…}}}, 4: {0: {9: {,…}}},…}
1: {0: {9: {,…}}}
2: {0: {9: {,…}}}
3: {0: {9: {,…}}}
4: {0: {9: {,…}}}
5: {0: {9: {d: [[1185], [1186], [1187], [1188], [1189]]}}}
6: {0: {9: {d: [[1190], [1191], [1192], [1193], [1194], [1195], [1196], [1197]]}}}
7: {0: {9: {d: [[1198], [1199], [1200], [1201], [1202], [1203], [1204]]}}}
8: {0: {9: {d: [[1205], [1206]]}}}
9: {0: {9: {d: [[1207], [1208], [1209], [1210], [1211]]}}}

Without it =>
diff: {3: {,…}, 4: {}, 5: {}, 9: {}, 10: {d: [[1], [2], [3], [4], [5], [6], [7], [8], [9]]}, 11: "",…}
3: {,…}
4: {}
5: {}
9: {}
10: {d: [[1], [2], [3], [4], [5], [6], [7], [8], [9]]}
11: ""
c: {1: {0: {4: {}, 6: {0: {}, 1: {}}, 8: {}, 9: {,…}, 10: {}}},…}
1: {0: {4: {}, 6: {0: {}, 1: {}}, 8: {}, 9: {,…}, 10: {}}}
2: {0: {4: {}, 6: {0: {}, 1: {}}, 8: {}, 9: {,…}, 10: {}}}
3: {0: {4: {}, 6: {0: {}, 1: {}}, 8: {}, 9: {,…}, 10: {}}}
4: {0: {4: {}, 6: {0: {}, 1: {}}, 8: {}, 9: {,…}, 10: {}}}
5: {0: {4: {}, 6: {0: {}, 1: {}}, 8: {}, 9: {d: [[1186], [1187], [1188], [1189], [1190]]}, 10: {}}}
6: {0: {4: {}, 6: {0: {}, 1: {}}, 8: {},…}}
7: {,…}
8: {0: {4: {}, 6: {0: {}, 1: {}}, 8: {}, 9: {d: [[1206], [1207]]}, 10: {}}}
9: {0: {4: {}, 6: {0: {}, 1: {}}, 8: {}, 9: {d: [[1208], [1209], [1210], [1211], [1212]]}, 10: {}}}
10: {1: {}, 3: {}}
11: {1: {}, 3: {}}
12: {1: {}, 3: {}}
13: {1: {}, 3: {}}
14: {1: {}, 3: {}}
15: {1: {}, 3: {}}
16: {1: {}, 3: {}}
17: {1: {}, 3: {}}
18: {1: {}, 3: {}}
19: {1: {}, 3: {}}
20: {1: {}, 3: {}}
...
1205: {1: {}, 3: {}}
1206: {1: {}, 3: {}}
1207: {1: {}, 3: {}}
1208: {1: {}, 3: {}}
1209: {1: {}, 3: {}}
1210: {1: {}, 3: {}}
1211: {1: {}, 3: {}}
1212: {1: {}, 3: {}}

**Comment by josevalim** [comment]
I think it is best if we discuss these optimizations one by one then, and compare the diffs before and after one by one. I have sent a separate PR for only the component one here: https://github.com/phoenixframework/phoenix_live_view/pull/898

Can you please make this PR only about the traverse_dynamic and send the diffs side by side? Thanks.

**Comment by seb3s** [comment]
| Without Optimization | With Optimization |
| -------------------- | ----------------- |
| diff: {3: {,…}, 4: {}, 5: {}, 9: {}, 10: {d: [[1], [2], [3], [4], [5], [6], [7], [8], [9]]}, 11: "",…} | diff: {3: {,…}, 10: {d: [[1], [2], [3], [4], [5], [6], [7], [8], [9]]}, 11: "",…} |
| 3: {,…} | 3: {,…} |
| 4: {} | |
| 5: {} | |
| 9: {} | |
| 10: {d: [[1], [2], [3], [4], [5], [6], [7], [8], [9]]} | 10: {d: [[1], [2], [3], [4], [5], [6], [7], [8], [9]]} |
| 11: "" | 11: "" |
| c: {1: {0: {4: {}, 6: {0: {}, 1: {}}, 8: {}, 9: {,…}, 10: {}}},…} | c: {1: {0: {9: {,…}}}, 2: {0: {9: {,…}}}, 3: {0: {9: {,…}}}, 4: {0: {9: {,…}}},…} |
| 1: {0: {4: {}, 6: {0: {}, 1: {}}, 8: {}, 9: {,…}, 10: {}}} | 1: {0: {9: {,…}}} |
| 2: {0: {4: {}, 6: {0: {}, 1: {}}, 8: {}, 9: {,…}, 10: {}}} | 2: {0: {9: {,…}}} |
| 3: {0: {4: {}, 6: {0: {}, 1: {}}, 8: {}, 9: {,…}, 10: {}}} | 3: {0: {9: {,…}}} |
| 4: {0: {4: {}, 6: {0: {}, 1: {}}, 8: {}, 9: {,…}, 10: {}}} | 4: {0: {9: {,…}}} |
| 5: {0: {4: {}, 6: {0: {}, 1: {}}, 8: {}, 9: {d: [[1186], [1187], [1188], [1189], [1190]]}, 10: {}}} | 5: {0: {9: {d: [[1186], [1187], [1188], [1189], [1190]]}}} |
| 6: {0: {4: {}, 6: {0: {}, 1: {}}, 8: {},…}} | 6: {0: {9: {d: [[1191], [1192], [1193], [1194], [1195], [1196], [1197], [1198]]}}} |
| 7: {0: {4: {}, 6: {0: {}, 1: {}}, 8: {}, 9: {d: [[1199], [1200], [1201], [1202], [1203], [1204], [1205]]},…}} | 7: {0: {9: {d: [[1199], [1200], [1201], [1202], [1203], [1204], [1205]]}}} |
| 8: {0: {4: {}, 6: {0: {}, 1: {}}, 8: {}, 9: {d: [[1206], [1207]]}, 10: {}}} | 8: {0: {9: {d: [[1206], [1207]]}}} |
| 9: {0: {4: {}, 6: {0: {}, 1: {}}, 8: {}, 9: {d: [[1208], [1209], [1210], [1211], [1212]]}, 10: {}}} | 9: {0: {9: {d: [[1208], [1209], [1210], [1211], [1212]]}}} |
| 10: {1: {}, 3: {}} | |
| 11: {1: {}, 3: {}} | |
| 12: {1: {}, 3: {}} | |
| 13: {1: {}, 3: {}} | |
| 14: {1: {}, 3: {}} | |
| 15: {1: {}, 3: {}} | |
| ... | |
| 1207: {1: {}, 3: {}} | |
| 1208: {1: {}, 3: {}} | |
| 1209: {1: {}, 3: {}} | |
| 1210: {1: {}, 3: {}} | |
| 1211: {1: {}, 3: {}} | |
| 1212: {1: {}, 3: {}} | |

**Comment by seb3s** [comment]
I hope this is what you mean by side by side ;-)

**Comment by josevalim** [comment]
Yes, this is much easier, awesome work!

I believe I can see the cases where this optimization would apply: it seems to be a comprehension of components... where the components stay the same? I believe the next step is to write a test in diff_test.exs that shows the optimization in effect and we also avoid regressions in the future.

**Comment by seb3s** [comment]
Ouch, 
I've spent way too many hours trying to figure out how all these tests run and so on. I'm still very novice on that part ! And I think I've hit the wall.
Here is what I've done :
```
  defmodule RenderNestedComponent do
    use Phoenix.LiveComponent

    def render(assigns) do
      ~L"""
      <div>
      RENDER NESTED <%= @from %>
      <%= for tag <- @tags do %>
        <%= if @tag == tag do tag end %>
      <% end %>
      </div>
      """
    end
  end

  defmodule NestedRenderOnlyComponent do
    use Phoenix.LiveComponent

    def render(assigns) do
      ~L"""
      <div>
      <%= for val <- 1..3 do %>
        <%= live_component @socket, RenderNestedComponent, id: "#{@id}_#{val}", from: @from, tag: @tag, tags: @tags %>
      <% end %>
      </div>
      """
    end
  end

  describe "seb stateful components" do
    test "seb multiple render only" do
      tags = ["foo", "bar"]
      components = [
        %Component{id: "index_0", assigns: %{id: "index_0", from: :component1, tag: "foo", tags: tags}, component: NestedRenderOnlyComponent},
        %Component{id: "index_1", assigns: %{id: "index_1", from: :component2, tag: "bar", tags: tags}, component: NestedRenderOnlyComponent}
      ]
      assigns = %{components: components}
      rendered = ~L"""
        <%= for component <- @components do %>
          <%= component %>
        <% end %>
        """

      {previous_socket, _, previous_components} = render(rendered)
      #{socket, full_render, components} = render(rendered)

      tags = ["foo", "bar"]
      components = [
        %Component{id: "index_0", assigns: %{id: "index_0", from: :component1, tag: "foo", tags: tags}, component: NestedRenderOnlyComponent},
        %Component{id: "index_1", assigns: %{id: "index_1", from: :component2, tag: "bar", tags: tags}, component: NestedRenderOnlyComponent}
      ]
      assigns = %{components: components}
      rendered = ~L"""
        <%= for component <- @components do %>
          <%= component %>
        <% end %>
        """
      {socket, full_render, components} =
        render(rendered, previous_socket.fingerprints, previous_components)

      assert full_render == %{
        0 => %{}
      }

      assert socket.fingerprints != {rendered.fingerprint, %{}}
      assert components == Diff.new_components()
    end
  end

```
I've tried various combinations, but i'm not able to submit a new assigns for rendering with only modified objects, it re-renders everything as It was submitted the first time, despite components shared the same ids. I'm definitely missing something.

Please enlighten me ;-)

**Comment by josevalim** [comment]
Thanks. I will take a look and try to take it from here. :)

**Comment by josevalim** [comment]
I have verified the patch is good. I will write a test shortly. :)

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**